cs-verifiable-encryption/README.md

```diff
@@ -0,0 +1,135 @@
+    ```
+
+1. To decrypt the ciphertext from above, the auditor will use the label, ciphertext and his keys in `CSKeypair::decrypt`. If successful, it returns the decrypted attributes
+
+    ```rust
+    let decrypted_messages: Vec<BigNumber> = CSKeypair::decrypt(label, &ciphertext, &keypair.pub_key, &keypair.pri_key)?;
+    ```
+
+1. The prover can now execute the verifiable encryption protocol. He will first either create or retrieve `blindings` for the attributes he wants to encrypt and then create the ciphertext over the attributes, over the `blindings` using `CSKeypair::encrypt_and_prove_phase_1`. He also gets some random values that he will use while creating the response to the challenge.
+
+    ```rust
+    let (ciphertext, blindings_ciphertext, r, r_tilde) = CSKeypair::encrypt_and_prove_phase_1(
+            &attributes,
+            &blindings,
+            label,
+            &keypair.pub_key,
+        )?;
+    ```    
+
+    The prover sends both `ciphertext` and `blindings_ciphertext` to the verifier. Incase of a non-interactive ZKP, both `ciphertext` and `blindings_ciphertext` should be hashed in the challenge.
+
+1. When the prover receives the challenge, it creates the response to it using `CSKeypair::encrypt_and_prove_phase_2`. The challenge could have come from a verifier in case of an interactive ZKP or it could be the result of hashing all the commitments of the main proving protocol (including this one).
+
+    ```rust
+    let r_hat: BigNumber = CSKeypair::encrypt_and_prove_phase_2(
+            &r,
+            &r_tilde,
+            &challenge,
+            &keypair.pub_key,
+            Some(&mut ctx),
+        )?;
+    ```    
+
+1. The verifier reconstructs the ciphertext over blindings and matches the received ciphertext over blindings using `CSKeypair::reconstruct_blindings_ciphertext`. The `attribute_responses` in the example below is sent by the prover as part of the main protocol for all attributes. The verifier picks those that were encrypted.
+
+    ```rust
+    let blindings_ciphertext: CSCiphertext = CSKeypair::reconstruct_blindings_ciphertext(
+            &ciphertext,
+            &attribute_responses,
+            &r_hat,
+            &challenge,
+            label,
+            &keypair.pub_key,
+        )
+    ```
+
+1. The private key `CSEncPrikey`, public key `CSEncPubkey` and ciphertext `CSCiphertext` are serializable using serde-json.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The prover should ensure that the auditor will not decrypt his attributes unless the pre-agreed terms are satisfied. If the verifier tricks the prover to encrypt his attributes for a colluding auditor then the prover's privacy is broken. Also the prover should be careful about what attributes he is encrypting and not encrypt sensitive attributes like link secret. The sensitivity of same attributes maybe different depending on the auditor. 
```

Comment:
Can we modify the implementation of this algorithm, either in Ursa or (more likely) in libindy, such that we guarantee that the link secret is never one of the values added to a VE block?

Reply:
That would be a specific implementation for Indy. This module is meant to be implementation agnostic. That sounds like a good idea.

Reply:
Updated text. I don't see why someone verifier will want the ciphertext over  link secret but don't know if that is a good enough reason to prevent it completely. What if Aries had 2 methods, one of them an unsafe variant which  allowed encrypting everything?